### PR TITLE
add new feature: Projectiles

### PR DESCRIPTION
**Pull request by t-m-z**
for review and discussion;

### CODE REVIEW

File: lib/widgets/tools/science_and_technology/projectiles.dart [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+            title: i18n(context, 'projectiles_mass'),
+            min: 0.0,
+            value: _currentMass,
+            onChanged: (value) {
+              setState(() {
+                _currentMass = value;
+              });
+            }
+        ),
+        GCWDoubleSpinner(
```
Comment (S-Man42) [inline lib/widgets/tools/science_and_technology/projectiles.dart:46]: Maybe it makes sense, to offer different units for speed and mass; as it was done in Windchill (https://github.com/S-Man42/GCWizard/blob/1.1.0/lib/widgets/tools/science_and_technology/windchill.dart) or even better in Beaufort (https://github.com/S-Man42/GCWizard/blob/1.1.0/lib/widgets/tools/science_and_technology/beaufort.dart) using the internal Unit Converter.

File: lib/widgets/tools/science_and_technology/projectiles.dart [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    String result = '';
+
+    if (_currentEnergy == 0) {
+      calculate = 'projectiles_energy';
+      result = calculateEnergy(_currentMass, _currentSpeed).toStringAsFixed(3);
+    } else
+      if (_currentMass == 0) {
+        calculate = 'projectiles_mass';
+        result = calculateMass(_currentEnergy, _currentSpeed).toStringAsFixed(3);
+      } else
```
Comment (S-Man42) [inline lib/widgets/tools/science_and_technology/projectiles.dart:72]: Just a minor issue, more something religious: I prefer to add the curly braces even around blocks like this. This increases readability and makes changes much easier.

File: lib/widgets/tools/science_and_technology/projectiles.dart [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+        if (_currentSpeed == 0) {
+          calculate = 'projectiles_speed';
+          result = calculateSpeed(_currentEnergy, _currentMass).toStringAsFixed(3);
+        } else {
+          calculate = 'projectiles_error';
+          result = '0';
+        }
+
+    return GCWOutput(
+      child: Column(
```
Comment (S-Man42) [inline lib/widgets/tools/science_and_technology/projectiles.dart:82]: Please check this with the refactored output system. maybe this results in missing "Output:" divider, I am not quite sure...

With the new system, i'd change it to:
```
GCWDefaultOutput(
    text: i18n(context, calculate) + ': ' + result,
   copyText: result
)
```

**Review by S-Man42 — COMMENTED** [review]
I confess, I didn't check this out so far, so I am not quite sure how this works. But it seems, you have 3 double spinners and if one of them equals 0, this is unit, you calculate, correct?

Well, if so, I'd change it as follows:
I would take a DropDown were the user can chose the unit, he wants to calculate and adjust the visible spinners (or just their labels...). This is much less "magic". Maybe the user wants to calculate with zero speed (and expects 0 energy ;)) It's not wrong anyway.

Furthermore, I'd add unit spinners. Using the internal Unit converter engine, this could be extremely powerful. Additionally this could be done with the output value as well. Have a look at the beaufort tool. Maybe we could extract new "unit widgets". The whole input widget for beaufort could be re-used as "velocity" input. For other unit, this could be adapted. An then, any further calculations can use them without worrying how to do this...

**Comment by S-Man42** [comment]
This PR has conflicts. Please resolve them.

**Comment by t-m-z** [comment]
> This PR has conflicts. Please resolve them.

solved

File: lib/widgets/registry.dart [hunk 4]
```diff
@@ -399,6 +400,12 @@ class Registry {
         category: ToolCategory.CRYPTOGRAPHY,
         searchStrings: 'geocaching geocache code gccode gc-code base31 hexadecimal hexadezimal'
       ),
+      GCWToolWidget(
+        tool: Projectiles(),
+        i18nPrefix: 'projectiles',
+        category: ToolCategory.SCIENCE_AND_TECHNOLOGY,
+        searchStrings: 'geschoss bullet projectile ballistik ballistic'
```
[the remark this answers is not in the document]
Comment (t-m-z) [reply lib/widgets/registry.dart:408]: done

File: assets/i18n/en.json [hunk 5]
```diff
@@ -1231,6 +1231,24 @@
   "primes_integerfactorization_description" : "Returns the prime factorization of a number",
   "primes_integerfactorization_example" : "42 = 2 × 3 × 7",
 
+  "projectiles_title" : "Bullets/projectiles",
+  "projectiles_description" : "calculate kinetic energy, mass and speed of bullets/projectiles",
```
[the remark this answers is not in the document]
Comment (t-m-z) [reply assets/i18n/en.json:1253]: done

File: assets/i18n/de.json [hunk 6]
```diff
@@ -1231,6 +1231,24 @@
...
+  "projectiles_title" : "Geschosse",
+  "projectiles_description" : "Berechnet kinetische Energie, Masse und Geschwindigkeit von projectilesen",
+  "projectiles_energy" : "kinetische Energie [Joule]",
+  "projectiles_mass" : "Masse [kg]",
+  "projectiles_speed" : "Geschwindigkeit [m/s]",
+  "projectiles_calculate" : "berechnen",
+  "projectiles_error_noInput_Energy" : "fehlende Eingabe: Energie [Joule]",
+  "projectiles_error_noInput_Mass" : "fehlende Eingabe: Masse [kg]",
+  "projectiles_error_noInput_Speed" : "fehlende Eingabe: Geschwindigkeit [m/s]",
+  "projectiles_error_wrongFormat_Double_Energy" : "Eingabe für Energie ist keine korrekte Reelle Zahl",
```
[the remark this answers is not in the document]
Comment (t-m-z) [reply assets/i18n/de.json:1261]: done

File: assets/i18n/de.json [hunk 7]
```diff
@@ -1231,6 +1231,24 @@
   "primes_integerfactorization_description" : "Primfaktorzerlegung einer Zahl",
   "primes_integerfactorization_example" : "42 = 2 × 3 × 7",
 
+  "projectiles_title" : "Geschosse",
+  "projectiles_description" : "Berechnet kinetische Energie, Masse und Geschwindigkeit von projectilesen",
+  "projectiles_energy" : "kinetische Energie [Joule]",
```
[the remark this answers is not in the document]
Comment (t-m-z) [reply assets/i18n/de.json:1254]: done

File: assets/i18n/de.json [hunk 8]
```diff
@@ -1231,6 +1231,24 @@
   "primes_integerfactorization_description" : "Primfaktorzerlegung einer Zahl",
   "primes_integerfactorization_example" : "42 = 2 × 3 × 7",
 
+  "projectiles_title" : "Geschosse",
+  "projectiles_description" : "Berechnet kinetische Energie, Masse und Geschwindigkeit von projectilesen",
```
[the remark this answers is not in the document]
Comment (t-m-z) [reply assets/i18n/de.json:1253]: Done

File: lib/widgets/tools/science_and_technology/projectiles.dart [hunk 9]
```diff
@@ -0,0 +1,381 @@
...
+        ),
+
+        _currentCalculateMode == CalculateProjectilesMode.ENERGY
+          ? Column(
+            children: <Widget>[
+              Row(
+                children: [
+                  Expanded(
+                    child: Container(
+                      child: GCWDoubleSpinner(
```
Comment (S-Man42) [inline lib/widgets/tools/science_and_technology/projectiles.dart:133]: This whole widgets is repeating as well: `Row(DoubleSpinner, <YourUnit>DropDownButton)`. This should be refactored as well. So, in fact, we need two new Widgets:` UnitDropDownButton(unit: ...)` and `UnitInput(unit: ...)`

Internally, UnitInput is `Row(DoubleSpinner, UnitDropDownButton(unit: unit))`.

This then makes the current code much smaller:

```
_buildInputFields() {
switch(_currentCalculateMode) {
   case MASS: unit1 = Energy; unit2 = VELOCITY; break;
  case...
}

return Column(
 children: [
   UnitInput(unit: unit1),
  UnitInput(unit: unit2)
 ]
);
}
```

File: lib/logic/units/mass.dart [hunk 10]
```diff
@@ -21,6 +21,12 @@ final MASS_GRAM = Mass(
   isReference: true
 );
 
+final MASS_KGRAM = Mass(
```
[the remark this answers is not in the document]
Comment (t-m-z) [reply lib/logic/units/mass.dart]: Done

[the remark this answers is not in the document]
Comment (t-m-z) [reply lib/logic/units/mass.dart]: thanks for this hint - the Widget is rewritten and everything is working fine. Now I am heading for the next level.

File: lib/widgets/common/gcw_mass_dropdownbutton.dart [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+  Widget build(BuildContext context) {
+    return GCWDropDownButton(
+      value: widget.value ?? _currentMassUnit,
+      onChanged: (newValue) {
+        setState(() {
+          _currentMassUnit = newValue;
+          widget.onChanged(newValue);
+        });
+      },
+      items: logic.masses.map((length) {
```
[the remark this answers is not in the document]
Comment (t-m-z) [reply lib/widgets/common/gcw_mass_dropdownbutton.dart]: I see the point.
I tried to create a new widget: gcw_unit_dropdownbutton with a list as parameter. This widget is used by gcw_unit_input which is used in projectiles.dart. The hierarchy is gcw_dropdownbutton - gcw_unit_dropdownbutton - gcw_unit_input.

However, I receive an error message "type 'List<dynamic>' is not a subtype of type 'List<DropdownMenuItem<dynamic>>'" which is caused by gcw_unit_dropdownbutton in line 22.

File: lib/widgets/tools/science_and_technology/projectiles.dart (hunk 9 above)
Comment (t-m-z) [reply lib/widgets/tools/science_and_technology/projectiles.dart:186]: I see the point. However, I beleive this is not so easy. I think the code should look like 

                    switch (_currentCalculateMode){
                      case CalculateProjectilesMode.ENERGY:
                        titleInput1 = 'projectiles_mass';
                        titleInput2 = 'projectiles_velocity';
                        masses.forEach((element) { unit1.add(element);});
                        velocities.forEach((element) { unit2.add(element);});
                        break;


        GCWUnitInput(
          title: i18n(context, titleInput1),
          min: 0.0,
          numberDecimalDigits: 3,
          value: _currentInput1,
          unit: _currentUnit1,
          items: unit1,
          onValueChanged: (value) {
            setState(() {
              _currentInput1 = value;
            });
          },
          onUnitChanged: (unit) {
            setState(() {
              _currentUnit1 = unit;
            });
          },
        ),

As written above, I started a Top Down Approach but I am stuck.

**Comment by t-m-z** [comment]
I pulled your comits - the result is great!
I start now to write the tests

**Pull request closed without merging**